### PR TITLE
Resolve #1136: Implement Lucene IndexMaintainer and RecordCursor

### LINKED ISSUE #1136: Implement Lucene IndexMaintainer and RecordCursor with basic cursor support
Put basics in place for index update/delete and basic scans.  Add cursor support (skip, continuations, limit).

### PR DESCRIPTION
**Pull request by jleach4**
Supports updates/deletes with cursor support (limit, skip, continuations).  

The LuceneIndexMaintainer provides core functionality for scanning Lucene Indexes backed by FoundationDB.
The core scan interface is the existing <a href="https://lucene.apache.org/core/7_7_3/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#package.description"> Lucene Query Parser syntax</a>.

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/IndexTypes.java
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/IndexTypes.java:148]: Index types not defined in the core usually don't get named there, but rather in the .jar that adds them.

For instance, `fdb-record-layer-spatial/src/main/java/com/apple/foundationdb/record/spatial/geophile/GeophileIndexTypes`.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java:710]: I don't think this is the intended semantics. It look to me like, if any pending check is ready, check it and allow the named check to supersede what was there before, whatever state it was in. Which means if a pending check by the given name is not yet ready, the new check with that name is ignored.

Questions about the desired behavior:
* If a check by the given name is ready, is it checked or is that result ignored? Assuming the check passes, is the new check remembered as well?
* If a check by the given name is pending, does the new check replace it?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java:65]: Ideally, this would get an `Executor` from someplace. Could the check be instantiated by an `FDBRecordContext` instead?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java:81]: Probably safer to synchronize on the context, rather than the state, since that's where the named checks are synchronized.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java:53]: I kind of wish there were a better place for this to live. Like if a context had a scratchpad, it could do there and then separately a commit hook could close it down. On the other hand, pulling it out of the registered commit hook should work, so I guess it's okay.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:80]: This is a `TODO`, right?

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:109]: Should check for `BY_LUCENE_QUERY` and throw if not, I think.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:113]: Not for now and really a core issue, but seems like we're missing an abstraction in the index API. Like maybe there are index-supported table functions like there are record functions and aggregate functions. And you use one of those with a query argument rather than a scan.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:143]: Are these potentially blocking calls, too? Or do they just queue work and I/O happens at commit time? Mainly for the record; we don't need to fix absolutely everything all at once.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:212]: Why not? Doesn't saving the same record twice overwrite the document?

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (sophiabuell) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:88]: nitpick: this should be defined above if its not reliant on arguments.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (sophiabuell) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:144]: This seems to queue work as the change isn't visible until commit is called. Though according to some javadocs deeper in Lucene if the deletes are using too much heap they'll be written to disk. So... both?

**Comment by scgray** [comment]
I'm about 3/4 done reviewing and should be complete later today (so far mostly ignorable nit-picks).

**Review by scgray — CHANGES_REQUESTED** [review]
Mostly my comments were optional.

**Comment by alecgrieser** [comment]
Sorry had to get to a few other things. Will try and take a look tomorrow and/or Wednesday.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java
Comment (sophiabuell) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java:710]: In the current situation a new check with the same name would 
- if the check is ready it would be checked and returned then the new check is remembered
- If it is not ready the check is not replaced and the old one is remembered. 
I believe the solve here if we WANT to replace the old check would be to create a specific replaceCommitCheck function.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (sophiabuell) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:212]: yes it does overwrite the document.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java
Comment (jleach4) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java]: @MMcM Added a session concept for lookup...  Hope that clarifies.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:145]: But mostly happens without I/O, so maybe okay for now.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java:54]: The new `session` stuff is what I meant by "scratchpad."

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:115]: Not for now.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java
Comment (MMcM) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java:91]: Does `IndexWriter.close` need to happen on transaction rollback (that is, `close` without `commit`)?

That was the idea behind having the session objects be something narrower than `Object` so they could have a lifecycle. Maybe the Java GC does a good enough job?

**Review by MMcM — CHANGES_REQUESTED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java
Comment (jleach4) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/IndexWriterCommitCheckAsync.java:91]: It should be collected by GC.  

`    // Ensure that only one thread actually gets to do the
    // closing
    if (shouldClose(true)) {
      try {
        if (infoStream.isEnabled("IW")) {
          infoStream.message("IW", "now flush at close");
        }

        flush(true, true);
        waitForMerges();
        commitInternal(config.getMergePolicy());
        rollbackInternal(); // ie close, since we just committed
      } catch (Throwable t) {
        // Be certain to close the index on any exception
        try {
          rollbackInternal();
        } catch (Throwable t1) {
          t.addSuppressed(t1);
        }
        throw t;
      }
    }`

**Review by MMcM — APPROVED** [review]

**Review by scgray — APPROVED** [review]

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java:229]: I'm assuming this is to deal with this note from the [lucene javadoc](https://lucene.apache.org/core/7_7_2/core/org/apache/lucene/store/Directory.html)
> A file open for writing may not be available for read access until the corresponding IndexOutput is closed.

That being said, will this work correctly if the writer is created after starting the cursor. Eg you wanted to do something like an update, where you pipe the result into code that updates records (e.g. I want to update a typo in all the docs)

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:57]: There is another index for full text queries, I don't think there would actually be any confusion based on the name, but probably still better. 
```suggestion
 * Tests for {@code LUCENE} type indexes.
```

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:119]: It looks like `FORWARD_SCAN` is another thing ignored by the actual index maintainer / scan logic.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:140]: This manually sets the continuation to 2, I think it would be better to have it query look at 2 records, and then query again with the continuation, and make sure that you see all the appropriate entries.

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java
Comment (ScottDugas) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java:111]: These seem like a good start to the tests, but there are a few things I can think of to add (either now or later), mostly inspired by the bugs I think I saw in the code above:
1. Multiple indexes on the same store
2. Multiple stores with the same index in the same context
3. Complex queries that match a document on both fields, and use an OR
4. calling `scanIndexRecords` and getting the actual document back
5. saving some records, committing, querying and mapping that into an update
6. Adding to at least some of these tests assertions about the resulting index entries.
7. Updating the metadata to cause the index subpsace to change mid-transaction


(Note: I would not be surprised if that list of tests goes noticeably beyond the testing of many of the existing cursors/indexes, but in that case it's an issue there, not here.

**Review by ScottDugas — CHANGES_REQUESTED** [review]
There's quite a few things in this review, I think almost all of them could be pushed out into a followup issue, since this is experimental, except the session on the RecordContext, although if those are marked EXPERIMENTAL too, I think I'll leave it up to you whether to address the other comments now or put them in an issue. I'm not sure if we currently have a centralized issue tracking these things yet, but it would be good.

File: fdb-record-layer-lucene/docs/index.md
Comment (jleach4) [reply fdb-record-layer-lucene/docs/index.md:17]: Fixing...

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java
Comment (jleach4) [reply fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneIndexTest.java]: Fixing

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java
Comment (jleach4) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java]: Fixing

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java
Comment (jleach4) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java]: Fixing...

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java
Comment (jleach4) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/DirectoryCommitCheckAsync.java]: Fixing...

**Review by ScottDugas — APPROVED** [review]
Seems acceptable for an experimental API, although there are definitely some rough edges, known bugs, and missing test coverage.

**Review by sophiabuell — APPROVED** [review]

**Pull request merged**
